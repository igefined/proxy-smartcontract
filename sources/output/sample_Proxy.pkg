{"name":"Proxy","code":"te6ccgECIwEABgkAART/APSkE/S88sgLAQIBYgIDA5rQAdDTAwFxsKMB+kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiFRQUwNvBPhhAvhi2zxVFNs88uCCyPhDAcx/AcoAVUDbPMntVBoEBQIBIBITBPLtou37AZIwf+BwIddJwh+VMCDXCx/eIIIQjaDyx7qP0DDTHwGCEI2g8se68uCB1AHQAfpAASDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IgSbBJwgEAj2zwjVSB/VTBtbds8+EFvJBAjXwMC2zx/4CCCEHVTbzC6DA0OBgHOUFQg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIzxYSgQEBzwAhbrObfwHKAMhYzxbJAcyUcDLKAOJYIG6VMHABywGOHiDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IjPFuLIWBED5I61MNMfAYIQdVNvMLry4IH6QAEg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIMVVA2zxsFH/gIIIQlGqYtrqOqDDTHwGCEJRqmLa68uCB0z8BMcgBghCv+Q9XWMsfyz/J+EIBcG3bPH/gwACRMOMNcAkHCAE6bW0ibrOZWyBu8tCAbyIBkTLiECRwAwSAQlAj2zwNBNwg+QEggvBWHbkZHNoZIvr2Bm7STtxoFms5ML/rWbp/yERxLniPdbqPoVvbPCAgbvLQgHCAQCUgbvLQgNs8f1UwbW3bPAOkA3/bMeCC8G70GhYDJ5y9AzltOjWT0/RlyWhVQSWos5Bn6/yLBg1lugkMDQoAEvhCUlDHBfLghAN2jzAw+EFvJBAjXwMlgRFNAscF8vQhIG7y0IBwgEAlIG7y0IDbPH9VMG1t2zwDpAN/2zHgINdJwh/jAjAMDQsDRoAg1yFwgEAi2zwoVSB/VTBtbds8+EFvJBAjXwNSYts8f9sxDA0OAULIcAHLH28AAW+MbW+MAds8byIByZMhbrOWAW8iWczJ6DEPAcrIcQHKAVAHAcoAcAHKAlAFINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiM8WUAP6AnABymgjbrORf5MkbrPilzMzAXABygDjDSFus5x/AcoAASBu8tCAAcyVMXABygDiyQH7ABAADGwzA6RVIAC6INdKIddJlyDCACLCALGOSgNvIoB/Is8xqwKhBasCUVW2CCDCAJwgqgIV1xhQM88WQBTeWW8CU0GhwgCZyAFvAlBEoaoCjhIxM8IAmdQw0CDXSiHXSZJwIOLi6F8DAJh/AcoAyHABygBwAcoAJG6znX8BygAEIG7y0IBQBMyWNANwAcoA4iRus51/AcoABCBu8tCAUATMljQDcAHKAOJwAcoAAn8BygACyVjMAFggbpUwcAHLAY4eINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiM8W4skBzAIBIBQVAgEgHyACEbo0TbPNs8bFGBoWAgEgFxgAAiMCEbSju2ebZ42KMBoZAhG1STtnm2eNinAaGwACJAJ+7UTQ1AH4Y9IAAY6E2zxsFeD4KNcLCoMJuvLgifpAASDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IgB0ds8HB0ABlRyEAG++kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiAGBAQHXANIAAZPUAdCRbeIB+kAh1wsBwwCOHQEg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIkjFt4gHUAdAeAAhwbW1tAGD6QCHXCwHDAI4dASDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IiSMW3iMRUUQzAAubu9GCcFzsPV0srnsehOw51kqFG2aCcJ3WNS0rZHyzItOvLf3xYjmCcCBVwBuAZ2OUzlg6rkclssOCcJ2XTlqzTstzOg6WbZRm6KSCcEDOdWnnFfnSULAdYW4mR7KAIBSCEiABGwr7tRNDSAAGAAdbJu40NWlwZnM6Ly9RbVJLM3hGU0NYeDZlRG5BeFB6Z0NUN0h5UDRzdFI1RkVKUlRzV21YdERUYjN0gg","abi":"{\"name\":\"Proxy\",\"types\":[{\"name\":\"StateInit\",\"header\":null,\"fields\":[{\"name\":\"code\",\"type\":{\"kind\":\"simple\",\"type\":\"cell\",\"optional\":false}},{\"name\":\"data\",\"type\":{\"kind\":\"simple\",\"type\":\"cell\",\"optional\":false}}]},{\"name\":\"Context\",\"header\":null,\"fields\":[{\"name\":\"bounced\",\"type\":{\"kind\":\"simple\",\"type\":\"bool\",\"optional\":false}},{\"name\":\"sender\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}},{\"name\":\"value\",\"type\":{\"kind\":\"simple\",\"type\":\"int\",\"optional\":false,\"format\":257}},{\"name\":\"raw\",\"type\":{\"kind\":\"simple\",\"type\":\"slice\",\"optional\":false}}]},{\"name\":\"SendParameters\",\"header\":null,\"fields\":[{\"name\":\"bounce\",\"type\":{\"kind\":\"simple\",\"type\":\"bool\",\"optional\":false}},{\"name\":\"to\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}},{\"name\":\"value\",\"type\":{\"kind\":\"simple\",\"type\":\"int\",\"optional\":false,\"format\":257}},{\"name\":\"mode\",\"type\":{\"kind\":\"simple\",\"type\":\"int\",\"optional\":false,\"format\":257}},{\"name\":\"body\",\"type\":{\"kind\":\"simple\",\"type\":\"cell\",\"optional\":true}},{\"name\":\"code\",\"type\":{\"kind\":\"simple\",\"type\":\"cell\",\"optional\":true}},{\"name\":\"data\",\"type\":{\"kind\":\"simple\",\"type\":\"cell\",\"optional\":true}}]},{\"name\":\"ChangeOwner\",\"header\":2174598809,\"fields\":[{\"name\":\"queryId\",\"type\":{\"kind\":\"simple\",\"type\":\"uint\",\"optional\":false,\"format\":64}},{\"name\":\"newOwner\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}}]},{\"name\":\"ChangeOwnerOk\",\"header\":846932810,\"fields\":[{\"name\":\"queryId\",\"type\":{\"kind\":\"simple\",\"type\":\"uint\",\"optional\":false,\"format\":64}},{\"name\":\"newOwner\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}}]},{\"name\":\"Deploy\",\"header\":2490013878,\"fields\":[{\"name\":\"queryId\",\"type\":{\"kind\":\"simple\",\"type\":\"uint\",\"optional\":false,\"format\":64}}]},{\"name\":\"DeployOk\",\"header\":2952335191,\"fields\":[{\"name\":\"queryId\",\"type\":{\"kind\":\"simple\",\"type\":\"uint\",\"optional\":false,\"format\":64}}]},{\"name\":\"FactoryDeploy\",\"header\":1829761339,\"fields\":[{\"name\":\"queryId\",\"type\":{\"kind\":\"simple\",\"type\":\"uint\",\"optional\":false,\"format\":64}},{\"name\":\"cashback\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}}]},{\"name\":\"ProxyMessage\",\"header\":2376135367,\"fields\":[{\"name\":\"str\",\"type\":{\"kind\":\"simple\",\"type\":\"string\",\"optional\":false}},{\"name\":\"to\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}}]},{\"name\":\"ChangeMsgOwner\",\"header\":1968402224,\"fields\":[{\"name\":\"addr\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}}]},{\"name\":\"LastMessage\",\"header\":null,\"fields\":[{\"name\":\"message\",\"type\":{\"kind\":\"simple\",\"type\":\"string\",\"optional\":true}},{\"name\":\"sender\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":true}},{\"name\":\"receiver\",\"type\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":true}}]}],\"receivers\":[{\"receiver\":\"internal\",\"message\":{\"kind\":\"text\",\"text\":\"Repeat\"}},{\"receiver\":\"internal\",\"message\":{\"kind\":\"text\",\"text\":\"Return\"}},{\"receiver\":\"internal\",\"message\":{\"kind\":\"text\"}},{\"receiver\":\"internal\",\"message\":{\"kind\":\"typed\",\"type\":\"ProxyMessage\"}},{\"receiver\":\"internal\",\"message\":{\"kind\":\"typed\",\"type\":\"ChangeMsgOwner\"}},{\"receiver\":\"internal\",\"message\":{\"kind\":\"typed\",\"type\":\"Deploy\"}}],\"getters\":[{\"name\":\"owner\",\"arguments\":[],\"returnType\":{\"kind\":\"simple\",\"type\":\"address\",\"optional\":false}},{\"name\":\"get_count\",\"arguments\":[],\"returnType\":{\"kind\":\"simple\",\"type\":\"int\",\"optional\":false,\"format\":257}},{\"name\":\"get_last\",\"arguments\":[],\"returnType\":{\"kind\":\"simple\",\"type\":\"LastMessage\",\"optional\":false}}],\"errors\":{\"2\":{\"message\":\"Stack undeflow\"},\"3\":{\"message\":\"Stack overflow\"},\"4\":{\"message\":\"Integer overflow\"},\"5\":{\"message\":\"Integer out of expected range\"},\"6\":{\"message\":\"Invalid opcode\"},\"7\":{\"message\":\"Type check error\"},\"8\":{\"message\":\"Cell overflow\"},\"9\":{\"message\":\"Cell underflow\"},\"10\":{\"message\":\"Dictionary error\"},\"13\":{\"message\":\"Out of gas error\"},\"32\":{\"message\":\"Method ID not found\"},\"34\":{\"message\":\"Action is invalid or not supported\"},\"37\":{\"message\":\"Not enough TON\"},\"38\":{\"message\":\"Not enough extra-currencies\"},\"128\":{\"message\":\"Null reference exception\"},\"129\":{\"message\":\"Invalid serialization prefix\"},\"130\":{\"message\":\"Invalid incoming message\"},\"131\":{\"message\":\"Constraints error\"},\"132\":{\"message\":\"Access denied\"},\"133\":{\"message\":\"Contract stopped\"},\"134\":{\"message\":\"Invalid argument\"},\"135\":{\"message\":\"Code of a contract was not found\"},\"136\":{\"message\":\"Invalid address\"},\"137\":{\"message\":\"Masterchain support is not enabled for this contract\"},\"4429\":{\"message\":\"Invalid sender\"}},\"interfaces\":[\"org.ton.introspection.v0\",\"org.ton.abi.ipfs.v0\",\"org.ton.deploy.lazy.v0\",\"org.ton.chain.workchain.v0\",\"org.ton.ownable\"]}","init":{"kind":"direct","args":[{"name":"owner","type":{"kind":"simple","type":"address","optional":false}}],"prefix":{"bits":1,"value":0},"deployment":{"kind":"system-cell","system":"te6cckECJQEABhMAAQHAAQEFoMZ3AgEU/wD0pBP0vPLICwMCAWIEEwOa0AHQ0wMBcbCjAfpAASDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IhUUFMDbwT4YQL4Yts8VRTbPPLggsj4QwHMfwHKAFVA2zzJ7VQbBREE8u2i7fsBkjB/4HAh10nCH5UwINcLH94gghCNoPLHuo/QMNMfAYIQjaDyx7ry4IHUAdAB+kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiBJsEnCAQCPbPCNVIH9VMG1t2zz4QW8kECNfAwLbPH/gIIIQdVNvMLoMDhAGA+SOtTDTHwGCEHVTbzC68uCB+kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiDFVQNs8bBR/4CCCEJRqmLa6jqgw0x8BghCUapi2uvLggdM/ATHIAYIQr/kPV1jLH8s/yfhCAXBt2zx/4MAAkTDjDXAJBwgBOm1tIm6zmVsgbvLQgG8iAZEy4hAkcAMEgEJQI9s8DgTcIPkBIILwVh25GRzaGSL69gZu0k7caBZrOTC/61m6f8hEcS54j3W6j6Fb2zwgIG7y0IBwgEAlIG7y0IDbPH9VMG1t2zwDpAN/2zHggvBu9BoWAyecvQM5bTo1k9P0ZcloVUElqLOQZ+v8iwYNZboJDA4KABL4QlJQxwXy4IQDdo8wMPhBbyQQI18DJYERTQLHBfL0ISBu8tCAcIBAJSBu8tCA2zx/VTBtbds8A6QDf9sx4CDXScIf4wIwDA4LA0aAINchcIBAIts8KFUgf1UwbW3bPPhBbyQQI18DUmLbPH/bMQwOEAFCyHAByx9vAAFvjG1vjAHbPG8iAcmTIW6zlgFvIlnMyegxDQC6INdKIddJlyDCACLCALGOSgNvIoB/Is8xqwKhBasCUVW2CCDCAJwgqgIV1xhQM88WQBTeWW8CU0GhwgCZyAFvAlBEoaoCjhIxM8IAmdQw0CDXSiHXSZJwIOLi6F8DAcrIcQHKAVAHAcoAcAHKAlAFINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiM8WUAP6AnABymgjbrORf5MkbrPilzMzAXABygDjDSFus5x/AcoAASBu8tCAAcyVMXABygDiyQH7AA8AmH8BygDIcAHKAHABygAkbrOdfwHKAAQgbvLQgFAEzJY0A3ABygDiJG6znX8BygAEIG7y0IBQBMyWNANwAcoA4nABygACfwHKAALJWMwADGwzA6RVIAHOUFQg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIzxYSgQEBzwAhbrObfwHKAMhYzxbJAcyUcDLKAOJYIG6VMHABywGOHiDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IjPFuLIWBIAWCBulTBwAcsBjh4g10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIzxbiyQHMAgEgFCACASAVFwIRujRNs82zxsUYGxYAAiMCASAYGgIRtKO7Z5tnjYowGxkAAiQCEbVJO2ebZ42KcBsfAn7tRNDUAfhj0gABjoTbPGwV4Pgo1wsKgwm68uCJ+kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiAHR2zwcHgG++kABINdJgQELuvLgiCDXCwoggQT/uvLQiYMJuvLgiAGBAQHXANIAAZPUAdCRbeIB+kAh1wsBwwCOHQEg10mBAQu68uCIINcLCiCBBP+68tCJgwm68uCIkjFt4gHUAdAdAGD6QCHXCwHDAI4dASDXSYEBC7ry4Igg1wsKIIEE/7ry0ImDCbry4IiSMW3iMRUUQzAACHBtbW0ABlRyEAIBICEiALm7vRgnBc7D1dLK57HoTsOdZKhRtmgnCd1jUtK2R8syLTry398WI5gnAgVcAbgGdjlM5YOq5HJbLDgnCdl05as07LczoOlm2UZuikgnBAznVp5xX50lCwHWFuJkeygCAUgjJAARsK+7UTQ0gABgAHWybuNDVpcGZzOi8vUW1SSzN4RlNDWHg2ZURuQXhQemdDVDdIeVA0c3RSNUZFSlJUc1dtWHREVGIzdIICEz9TM="}},"sources":{"sources/contract.tact":"aW1wb3J0ICJAc3RkbGliL293bmFibGUiOwppbXBvcnQgIkBzdGRsaWIvZGVwbG95IjsKCm1lc3NhZ2UgUHJveHlNZXNzYWdlIHsKICAgIHN0cjogU3RyaW5nOwogICAgdG86IEFkZHJlc3M7Cn0KCm1lc3NhZ2UgQ2hhbmdlTXNnT3duZXIgewogICAgYWRkcjogQWRkcmVzczsKfQoKc3RydWN0IExhc3RNZXNzYWdlIHsKICAgIG1lc3NhZ2U6IFN0cmluZz87CiAgICBzZW5kZXI6IEFkZHJlc3M/OwogICAgcmVjZWl2ZXI6IEFkZHJlc3M/Owp9Cgp0cmFpdCBTdGF0cyB7CiAgICBjb3VudDogSW50OwogICAgbGFzdF9tZXNzYWdlOiBTdHJpbmc/OwogICAgbGFzdF9zZW5kZXI6IEFkZHJlc3M/OwogICAgbGFzdF9yZWNlaXZlcjogQWRkcmVzcz87CgogICAgZ2V0IGZ1biBnZXRfY291bnQoKTogSW50IHsKICAgICAgICByZXR1cm4gc2VsZi5jb3VudDsKICAgIH0KCiAgICBnZXQgZnVuIGdldF9sYXN0KCk6IExhc3RNZXNzYWdlIHsKICAgICAgICByZXR1cm4gTGFzdE1lc3NhZ2UgewogICAgICAgICAgICBtZXNzYWdlOiBzZWxmLmxhc3RfbWVzc2FnZSwKICAgICAgICAgICAgc2VuZGVyOiBzZWxmLmxhc3Rfc2VuZGVyLAogICAgICAgICAgICByZWNlaXZlcjogc2VsZi5sYXN0X3JlY2VpdmVyCiAgICAgICAgfTsKICAgIH0KfQoKY29udHJhY3QgUHJveHkgd2l0aCBPd25hYmxlLCBTdGF0cywgRGVwbG95YWJsZSB7CiAgICBvd25lcjogQWRkcmVzczsKCiAgICBjb3VudDogSW50ID0gMDsKICAgIGxhc3RfbWVzc2FnZTogU3RyaW5nPyA9IG51bGw7CiAgICBsYXN0X3NlbmRlcjogQWRkcmVzcz8gPSBudWxsOwogICAgbGFzdF9yZWNlaXZlcjogQWRkcmVzcz8gPSBudWxsOwoKICAgIGluaXQob3duZXI6IEFkZHJlc3MpIHsKICAgICAgICBzZWxmLm93bmVyID0gb3duZXI7CiAgICB9CgogICAgcmVjZWl2ZSgiUmVwZWF0IikgewogICAgICAgIHNlbGYucmVxdWlyZU93bmVyKCk7CgogICAgICAgIHNlbmQoU2VuZFBhcmFtZXRlcnN7CiAgICAgICAgICAgIHRvOiBzZWxmLmxhc3RfcmVjZWl2ZXIhISwKICAgICAgICAgICAgdmFsdWU6IDAsCiAgICAgICAgICAgIG1vZGU6IFNlbmRSZW1haW5pbmdWYWx1ZSwKICAgICAgICAgICAgYm9keTogKHNlbGYubGFzdF9tZXNzYWdlISEpLmFzQ29tbWVudCgpCiAgICAgICAgfSk7CgogICAgICAgIHNlbGYuY291bnQgPSBzZWxmLmNvdW50ICsgMTsKICAgIH0KCiAgICByZWNlaXZlKCJSZXR1cm4iKSB7CiAgICAgICAgbGV0IGN0eDogQ29udGV4dCA9IGNvbnRleHQoKTsKICAgICAgICByZXF1aXJlKGN0eC5zZW5kZXIgPT0gc2VsZi5vd25lciwgIkludmFsaWQgc2VuZGVyIik7CgogICAgICAgIHNlbmQoU2VuZFBhcmFtZXRlcnN7CiAgICAgICAgICAgIHRvOiBzZWxmLmxhc3Rfc2VuZGVyISEsCiAgICAgICAgICAgIHZhbHVlOiAwLAogICAgICAgICAgICBtb2RlOiBTZW5kUmVtYWluaW5nVmFsdWUsCiAgICAgICAgICAgIGJvZHk6IChzZWxmLmxhc3RfbWVzc2FnZSEhKS5hc0NvbW1lbnQoKQogICAgICAgIH0pOwoKICAgICAgICBzZWxmLmNvdW50ID0gc2VsZi5jb3VudCArIDE7CiAgICB9CgogICAgcmVjZWl2ZShtc2c6IFN0cmluZykgewogICAgICAgIHNlbmQoU2VuZFBhcmFtZXRlcnN7CiAgICAgICAgICAgIHRvOiBzZWxmLm93bmVyLAogICAgICAgICAgICB2YWx1ZTogMCwKICAgICAgICAgICAgbW9kZTogU2VuZFJlbWFpbmluZ1ZhbHVlLAogICAgICAgICAgICBib2R5OiBtc2cuYXNDb21tZW50KCkKICAgICAgICB9KTsKCiAgICAgICAgbGV0IGN0eDogQ29udGV4dCAgPSBjb250ZXh0KCk7CgogICAgICAgIHNlbGYuc3RvcmVTdGF0cyhjdHguc2VuZGVyLCBzZWxmLm93bmVyLCBtc2cpOwogICAgfQoKICAgIHJlY2VpdmUobXNnOiBQcm94eU1lc3NhZ2UpewogICAgICAgIHNlbmQoU2VuZFBhcmFtZXRlcnN7CiAgICAgICAgICAgIHRvOiBtc2cudG8sCiAgICAgICAgICAgIHZhbHVlOiAwLAogICAgICAgICAgICBtb2RlOiBTZW5kUmVtYWluaW5nVmFsdWUsCiAgICAgICAgICAgIGJvZHk6IG1zZy5zdHIuYXNDb21tZW50KCkKICAgICAgICB9KTsKCiAgICAgICAgbGV0IGN0eDogQ29udGV4dCA9IGNvbnRleHQoKTsKCiAgICAgICAgc2VsZi5zdG9yZVN0YXRzKGN0eC5zZW5kZXIsIG1zZy50bywgbXNnLnN0cik7CiAgICB9CgogICAgcmVjZWl2ZShtc2c6IENoYW5nZU1zZ093bmVyKSB7CiAgICAgICAgc2VsZi5yZXF1aXJlT3duZXIoKTsKICAgICAgICBzZWxmLm93bmVyID0gbXNnLmFkZHI7CiAgICB9CgogICAgZnVuIHN0b3JlU3RhdHMoc2VuZGVyOiBBZGRyZXNzLCByZWNlaXZlcjogQWRkcmVzcywgbXNnOiBTdHJpbmcpIHsKICAgICAgICBzZWxmLmNvdW50ID0gc2VsZi5jb3VudCArIDE7CiAgICAgICAgc2VsZi5sYXN0X3NlbmRlciA9IHNlbmRlcjsKICAgICAgICBzZWxmLmxhc3RfcmVjZWl2ZXIgPSByZWNlaXZlcjsKICAgICAgICBzZWxmLmxhc3RfbWVzc2FnZSA9IG1zZzsKICAgIH0KfQ=="},"compiler":{"name":"tact","version":"1.3.0","parameters":"{\"entrypoint\":\"./sources/contract.tact\",\"options\":{}}"}}