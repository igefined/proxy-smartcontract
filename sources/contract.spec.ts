import {toNano} from "@ton/core";
import {ContractSystem} from "@tact-lang/emulator";
import {Proxy} from "./output/sample_Proxy";
import {inspect} from "util";

describe("contract", () => {
    it("forward message", async () => {
        // Create ContractSystem and deploy contract
        let system = await ContractSystem.create();
        let owner = system.treasure("owner");
        let u1 = system.treasure("user_1");
        let u2 = system.treasure("user_2");
        let proxy = system.open(await Proxy.fromInit(owner.address));
        let proxyTracker = system.track(proxy)

        await proxy.send(u1, {value: toNano("0.1")}, {$$type: 'ProxyMessage', str: 'Hello', to: u2.address});
        await system.run();

        expect(await proxy.getGetCount()).toEqual(1n);
        expect((await proxy.getGetLast()).message).toEqual("Hello");
        expect(((await proxy.getGetLast()).receiver)?.equals(u2.address)).toBeTruthy();
        expect(((await proxy.getGetLast()).sender)?.equals(u1.address)).toBeTruthy();
        let proxyEvents = proxyTracker.collect();
        expect(proxyEvents).toMatchSnapshot();
        console.log(inspect(proxyEvents, true, null, true));
    });
});
