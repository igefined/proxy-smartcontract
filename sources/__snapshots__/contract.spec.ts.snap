// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contract forward message 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "cell": "x{8DA0F2C780083BDB46227A42A96323F29EA02287D2DC51FEE3B85A52EEE5534F7FB0671BCED_}
 x{48656C6C6F}",
            "type": "cell",
          },
          "bounce": true,
          "from": "@treasure(user_1)",
          "to": "kQA0fDcbiVa0MaEU9ZccI628l7hjZ6kCP-VcZB3Dp0BmlOrT",
          "type": "internal",
          "value": "0.1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 13051n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "text": "Hello",
              "type": "text",
            },
            "bounce": true,
            "from": "kQA0fDcbiVa0MaEU9ZccI628l7hjZ6kCP-VcZB3Dp0BmlOrT",
            "to": "@treasure(user_2)",
            "type": "internal",
            "value": "0.085777",
          },
        ],
      },
    ],
  },
  {
    "$seq": 1,
    "events": [
      {
        "$type": "received-bounced",
        "message": {
          "body": {
            "cell": "x{FFFFFFFF0000000048656C6C6F}",
            "type": "cell",
          },
          "bounce": false,
          "from": "@treasure(user_2)",
          "to": "kQA0fDcbiVa0MaEU9ZccI628l7hjZ6kCP-VcZB3Dp0BmlOrT",
          "type": "internal",
          "value": "0.084777",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8088n,
      },
    ],
  },
]
`;
